src/main/java/com/gmail/val59000mc/scenarios/scenariolisteners/RandomizedBlockDropsListener.java

```diff
@@ -0,0 +1,63 @@
+package com.gmail.val59000mc.scenarios.scenariolisteners;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.block.BlockBreakEvent;
+import org.bukkit.inventory.ItemStack;
+
+import com.gmail.val59000mc.scenarios.ScenarioListener;
+import com.gmail.val59000mc.utils.RandomUtils;
+
+public class RandomizedBlockDropsListener extends ScenarioListener{
+	
+	private List<Material> materialitems;
+	private HashMap<Material, ItemStack> dropList; 
+	
+	public RandomizedBlockDropsListener() {
+		
+		materialitems = new ArrayList<>();
+		dropList = new HashMap<Material, ItemStack>();
+	}
+
+	@Override
+	public void onEnable() {
+		//Create new arraylist of materials that are all items
+	for(int i = 0; i< Material.values().length; i++) {
+		if(Material.values()[i].isItem()) {
+			 materialitems.add(Material.values()[i]);
+			}
+		}
+	}
+	
+	@EventHandler
+	public void onBlockBreak(BlockBreakEvent event) {
+		
+		//Create new HashMap so each each type of broken block drops the same random item every time it is broken (configurable
+		Block block = event.getBlock();
+		int itemindex  = RandomUtils.randomInteger(1, materialitems.size())-1;
+		
+		ItemStack randomdrops = new ItemStack(materialitems.get(itemindex), 1);
+		
+		ItemStack itemdrops = dropList.get(block.getType());
+		
+		System.out.println(dropList);
+		if(dropList.containsKey(block.getType())) {
+			block.setType(Material.AIR);
+			block.getWorld().dropItemNaturally(block.getLocation(), itemdrops);
+			event.setCancelled(true);
+		}
+		else {
+			dropList.put(block.getType(), randomdrops);
+			block.setType(Material.AIR);
+			block.getWorld().dropItemNaturally(block.getLocation(), randomdrops);
+			event.setCancelled(true);
+		}
+	}
+}
```

Comment:
Lots of duplicate code.
"itemdrops" is instantiate without checking if the drop list contains the block type.
You forgot to remove the printLn. (I asume it was a debug message.)
Maybe when you add a item to the "dropList" remove it from the "materialItems" list to prevent two blocks from dropping the same item.
`block.getLocation()` returns a none decimal location (e.x. 30.0, 19.0, 863.0) at the corner of a block, this can couse items to spawn inside other blocks. To prevent this use the "add(.5, 0, .5)" method on "Location".

Reply:
While `itemdrops` is instantiated without checking if the `droplist` contains the block type, the next code block (the if statement) checks this. the `else` statement will drop `randomitems`, rather than the entry in the hashmap (which doesnt exist).

Reply:
Yea, but that still means your creating an instance of an object that you don't always use.

Reply:
How should I do it then? If I do `block.getWorld().dropItemNaturally(block.getLocation(), dropList.get(block.getType()));` I get an NPE at that line saying that I cannot drop a null item :(

Reply:
This is how I'd do it:
`	public void onBlockBreak(BlockBreakEvent event) {
		//Create new HashMap so each each type of broken block drops the same random item every time it is broken (configurable
		Block block = event.getBlock();
		ItemStack blockDrop;

		if(dropList.containsKey(block.getType())) {
			blockDrop = dropList.get(block.getType());
		}
		else {
			int itemindex  = RandomUtils.randomInteger(1, materialItems.size())-1;
			Material material = materialItems.get(itemIndex);

			blockDrop = new ItemStack(material);
			dropList.put(block.getType(), blockDrop);

			materialItems.remove(material);
		}

		event.setCancelled(true);
		block.setType(Material.AIR);
		Location dropLocation = block.getLocation().add(.5, 0, .5);
		dropLocation.getWorld().dropItemNaturally(dropLocation, blockDrop);
	}`

Reply:
thanks so much for the tips!!!!